File: engine/lab/tests/test_events.py
"""Wire-format tests for the §5.1 event types."""

import json

from cloze_lab.scheduler.events import (
    BlockFinalized,
    BlockStarted,
    CommitItem,
    GenFinished,
    GenStarted,
    StepStats,
    TokensCommitted,
    TokensRevised,
    ReviseItem,
    WorkspaceReadout,
    WorkspaceReadoutItem,
    event_to_dict,
    to_jsonl_line,
    write_jsonl,
)


def test_wire_keys_match_design_5_1() -> None:
    line = to_jsonl_line(
        TokensCommitted(t=7, block=3, items=(CommitItem(pos=101, id=4821, conf=0.93),))
    )
    d = json.loads(line)
    assert d == {
        "t": 7,
        "type": "tokens_committed",
        "block": 3,
        "items": [{"pos": 101, "id": 4821, "conf": 0.93}],
    }


def test_every_event_type_serializes() -> None:
    events = [
        GenStarted(t=0, prompt_tokens=412, block_len=32, max_new=512),
        BlockStarted(t=0, block=3, span=(96, 128)),
        TokensCommitted(t=7, block=3, items=()),
        TokensRevised(t=9, block=3, items=(ReviseItem(pos=99, old=311, id=7, conf=0.71),)),
        StepStats(t=7, block=3, step=7, committed=21, remaining=11, ms=38.2, cache_hit=0.82),
        BlockFinalized(t=9, block=3, text=" the cache is refreshed", steps_used=9),
        GenFinished(t=9, reason="eos", new_tokens=487, wall_ms=5210.0, steps_total=134, tok_per_s=93.5),
        WorkspaceReadout(t=10, run_id="run_demo", token_index=1, token_text=" cat", layer=12, position=1,
                         top_readouts=(WorkspaceReadoutItem(label="uncertainty", score=0.62),),
                         entropy=0.41, provider="mock"),
    ]
    names = [event_to_dict(e)["type"] for e in events]
    assert names == [
        "gen_started",
        "block_started",
        "tokens_committed",
        "tokens_revised",
        "step_stats",
        "block_finalized",
        "gen_finished",
        "workspace_readout",
    ]
    for e in events:
        assert json.loads(to_jsonl_line(e))["t"] == e.t


def test_workspace_readout_payload_shape() -> None:
    d = event_to_dict(
        WorkspaceReadout(t=3, run_id="run_demo", token_index=2, token_text=" maybe", layer=12, position=2,
                         top_readouts=(
                             WorkspaceReadoutItem(label="uncertainty", score=0.71),
                             WorkspaceReadoutItem(label="hallucination_risk", score=0.43),
                         ),
                         entropy=0.52, provider="mock", provider_type="mock", readout_kind="risk")
    )
    assert d == {
        "t": 3,
        "type": "workspace_readout",
        "run_id": "run_demo",
        "token_index": 2,
        "token_text": " maybe",
        "layer": 12,
        "position": 2,
        "top_readouts": [
            {"label": "uncertainty", "score": 0.71},
            {"label": "hallucination_risk", "score": 0.43},
        ],
        "entropy": 0.52,
        "provider": "mock",
        "provider_type": "mock",
        "readout_kind": "risk",
    }


def test_write_jsonl_roundtrip(tmp_path) -> None:
    events = [
        GenStarted(t=0, prompt_tokens=2, block_len=0, max_new=4),
        BlockStarted(t=0, block=0, span=(2, 6)),
    ]
    path = tmp_path / "run.jsonl"
    write_jsonl(events, path)
    lines = path.read_text(encoding="utf-8").splitlines()
    assert [json.loads(l) for l in lines] == [json.loads(to_jsonl_line(e)) for e in events]

File: engine/core/tests/test_events.cpp
// test_events.cpp — the §5.1 event spine (DESIGN invariant 2), backend-free. Drives generate
// with the deterministic FakeAdapter and checks the event stream: shape (gen_started first,
// gen_finished last, one block_started/finalized per block), that the tokens_committed items
// reconstruct the board, that the live on_event callback sees the same stream as
// GenerateResult.events, and that JSONL serialization round-trips line-for-line.
#include "cloze/events.hpp"
#include "cloze/generate.hpp"
#include "fake_adapter.hpp"

#include <cassert>
#include <cstdio>
#include <stdexcept>
#include <string>
#include <variant>
#include <vector>

using namespace cloze;

// Active even in Release (asserts compile out under NDEBUG): a hard check that reports + aborts.
#define CHECK(cond) do { if (!(cond)) { std::fprintf(stderr, "CHECK failed: %s (line %d)\n", #cond, __LINE__); return 3; } } while (0)

namespace {
const std::vector<int> PROMPT = {1, 2, 3, 4, 5};  // p = 5

template <class T>
int count_of(const std::vector<Event>& evs) {
    int c = 0;
    for (const Event& e : evs)
        if (std::holds_alternative<T>(e)) ++c;
    return c;
}
}  // namespace

int main() {
    // --- Whole-sequence: 4 slots drain in 4 quota passes (1 commit/pass).
    {
        FakeAdapter fake(16, /*eos=*/-1, /*eos_at=*/-1);
        GenerateConfig cfg{/*max_new=*/4, /*steps=*/4, /*block_len=*/0, /*topk=*/-1};

        std::vector<Event> live;
        GenerateResult r = generate(fake, PROMPT, cfg, CacheConfig{}, nullptr,
                                    [&](const Event& e) { live.push_back(e); });
        const std::vector<Event>& evs = r.events;

        // Live callback saw exactly the collected stream.
        CHECK(live.size() == evs.size());

        // Shape: first gen_started, last gen_finished.
        CHECK(std::holds_alternative<GenStarted>(evs.front()));
        CHECK(std::holds_alternative<GenFinished>(evs.back()));
        const auto& gs = std::get<GenStarted>(evs.front());
        CHECK(gs.prompt_tokens == 5 && gs.max_new == 4 && gs.block_len == 0);
        const auto& gf = std::get<GenFinished>(evs.back());
        CHECK(gf.reason == "length" && gf.new_tokens == 4 && gf.steps_total == r.steps_total);

        // One block, four steps.
        CHECK(count_of<BlockStarted>(evs) == 1);
        CHECK(count_of<BlockFinalized>(evs) == 1);
        CHECK(count_of<StepStats>(evs) == r.steps_total);  // 4

        // The committed items reconstruct the generated board (f(p) = p left-to-right).
        std::vector<int> recon = PROMPT;
        recon.resize(PROMPT.size() + cfg.max_new, fake.config().mask_token_id);
        int total_committed = 0;
        for (const Event& e : evs)
            if (const auto* tc = std::get_if<TokensCommitted>(&e))
                for (const CommitItem& it : tc->items) { recon[it.pos] = it.id; ++total_committed; }
        CHECK(total_committed == 4);
        CHECK((std::vector<int>(recon.begin() + 5, recon.end()) == std::vector<int>{5, 6, 7, 8}));

        // JSONL round-trips: every line non-empty and carries its type tag; file has one line/event.
        for (const Event& e : evs) {
            const std::string line = to_jsonl_line(e);
            CHECK(!line.empty() && line.find("\"type\":") != std::string::npos);
        }
        const std::string path = "test_events_out.jsonl";
        CHECK(write_jsonl(evs, path));
        std::FILE* f = std::fopen(path.c_str(), "rb");
        CHECK(f);
        int lines = 0;
        for (int ch; (ch = std::fgetc(f)) != EOF;)
            if (ch == '\n') ++lines;
        std::fclose(f);
        std::remove(path.c_str());
        CHECK(lines == static_cast<int>(evs.size()));
    }

    // --- Block mode: two blocks -> two block_started/finalized, spans contiguous and ordered.
    {
        FakeAdapter fake(16, -1, -1);
        GenerateConfig cfg{/*max_new=*/4, /*steps=*/4, /*block_len=*/2, /*topk=*/-1};
        GenerateResult r = generate(fake, PROMPT, cfg);
        CHECK(count_of<BlockStarted>(r.events) == 2);
        CHECK(count_of<BlockFinalized>(r.events) == 2);
        std::vector<std::pair<int, int>> spans;
        for (const Event& e : r.events)
            if (const auto* bs = std::get_if<BlockStarted>(&e)) spans.push_back(bs->span);
        CHECK(spans.size() == 2);
        CHECK(spans[0] == std::make_pair(5, 7) && spans[1] == std::make_pair(7, 9));
    }

    // --- EOS: gen_finished reason "eos", and a tokens_committed carries the eos id.
    {
        FakeAdapter fake(16, /*eos=*/2, /*eos_at=*/6);
        GenerateConfig cfg{4, 4, 0, -1};
        GenerateResult r = generate(fake, PROMPT, cfg);
        CHECK(std::get<GenFinished>(r.events.back()).reason == "eos");
    }

    // --- Workspace Lens event: additive JSONL wire type for future providers.
    {
        WorkspaceReadout wr{/*t=*/3, /*run_id=*/"run_demo", /*token_index=*/1, /*token_text=*/" cat",
                            /*layer=*/12, /*position=*/1,
                            /*top_readouts=*/{{"uncertainty", 0.62}, {"hallucination_risk", 0.31}},
                            /*entropy=*/0.44, /*provider=*/"mock"};
        const std::string line = to_jsonl_line(Event{wr});
        CHECK(line.find("\"type\": \"workspace_readout\"") != std::string::npos);
        CHECK(line.find("\"run_id\": \"run_demo\"") != std::string::npos);
        CHECK(line.find("\"provider\": \"mock\"") != std::string::npos);
        CHECK(line.find("\"provider_type\": \"mock\"") != std::string::npos);
        CHECK(line.find("\"readout_kind\": \"risk\"") != std::string::npos);
        CHECK(line.find("\"top_readouts\":") != std::string::npos);
    }

    std::printf("test_events: all assertions passed\n");
    return 0;
}
